modules/etcd/etcd.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	// we finally need to re-apply all the etcd-specific options
+	clusterOpts = append(clusterOpts, withClusterOptions(etcdOpts))
+
+	return clusterOpts, nil
+}
+
+// configureCMD configures the etcd command line arguments, based on the settings provided,
+// in order to create a cluster or a single-node instance.
+func configureCMD(settings options) []string {
+	cmds := []string{"etcd"}
+
+	if len(settings.nodeNames) == 0 {
+		cmds = append(cmds, "--name=default")
+	} else {
+		clusterCmds := []string{
+			"--name=" + settings.nodeNames[settings.currentNode],
+			"--initial-advertise-peer-urls=" + scheme + "://" + settings.nodeNames[settings.currentNode] + ":" + peerPort,
+			"--advertise-client-urls=" + scheme + "://" + settings.nodeNames[settings.currentNode] + ":" + clientPort,
+			"--listen-peer-urls=" + scheme + "://0.0.0.0:" + peerPort,
+			"--listen-client-urls=" + scheme + "://0.0.0.0:" + clientPort,
+			"--initial-cluster-state=new",
+		}
+
+		clusterStateValues := make([]string, len(settings.nodeNames))
+		for i, node := range settings.nodeNames {
+			clusterStateValues[i] = node + "=" + scheme + "://" + node + ":" + peerPort
+		}
+		clusterCmds = append(clusterCmds, "--initial-cluster="+strings.Join(clusterStateValues, ","))
+
+		if settings.clusterToken != "" {
+			clusterCmds = append(clusterCmds, "--initial-cluster-token="+settings.clusterToken)
+		}
+
+		cmds = append(cmds, clusterCmds...)
+	}
+
+	if settings.mountDataDir {
+		cmds = append(cmds, "--data-dir="+dataDir)
+	}
+
+	cmds = append(cmds, settings.additionalArgs...)
+
+	return cmds
+}
+
+// ClientEndpoint returns the client endpoint for the etcd container, and an error if any.
+// For a cluster, it returns the client endpoint of the first node.
+func (c *EtcdContainer) ClientEndpoint(ctx context.Context) (string, error) {
+	var errs []error
+
+	host, err := c.Host(ctx)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("host: %w", err))
+	}
+
+	port, err := c.MappedPort(ctx, clientPort)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("mapped port: %w", err))
+	}
+
+	return fmt.Sprintf("http://%s:%s", host, port.Port()), errors.Join(errs...)
+}
+
+// ClientEndpoints returns the client endpoints for the etcd cluster.
+func (c *EtcdContainer) ClientEndpoints(ctx context.Context) ([]string, error) {
+	var endpoints []string
+
+	endpoint, err := c.ClientEndpoint(ctx)
+	if err != nil {
+		return endpoints, err
+	}
+	endpoints = append(endpoints, endpoint)
+
+	for _, node := range c.childNodes {
+		endpoint, err := node.ClientEndpoint(ctx)
+		if err != nil {
+			return endpoints, err
+		}
+		endpoints = append(endpoints, endpoint)
+	}
+
+	return endpoints, nil
+}
+
+// PeerEndpoint returns the peer endpoint for the etcd container, and an error if any.
+// For a cluster, it returns the peer endpoint of the first node.
+func (c *EtcdContainer) PeerEndpoint(ctx context.Context) (string, error) {
+	var errs []error
```

Comment:
suggestion: you don't want best effort here, it was only need for the resource clean up as you want to try to remove all resources, in this case if any request fails the URL is unusable